qscintilla: fix build on darwin

willcohen opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Per #71398, qscintilla is currently broken on darwin. 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

###### Mechanism of change
@lsix @peterhoeg 

**Reorganize `qscintilla` packages to make qt5 the default and move the legacy version to `qscintilla-qt4`**:

1. Moves current `qscintilla` to `qscintilla-qt4` and drops its support for both `qt5` and `darwin`.
2. Makes current `qscintilla-qt5` the default for `qscintilla` and migrates to `libsForQt5`.
3. Moves current `qscintilla` python package to `qscintilla-qt4`.

**Upgrading qscintilla version**:

4. Update `qscintilla` to 2.13.1
5. Convert `qscintilla-qt5` to use `pyproject.toml` and ensure build on `darwin`.
6. Adjust derivations using `qscintilla` and `qt4` to specify `qscintilla-qt4`

###### Previous issues before patch was ready for review

~~Trying to build the current version leads to issues the following build problem on Darwin:~~

```
$ nix-build $NIXPKGS -A python3Packages.qscintilla-qt5
                               
this derivation will be built:
  /nix/store/g3rv06mv76dg46gcfk80gnr89d111jhr-python3.9-qscintilla-2.11.6.drv
building '/nix/store/g3rv06mv76dg46gcfk80gnr89d111jhr-python3.9-qscintilla-2.11.6.drv'...
Sourcing python-remove-tests-dir-hook
Sourcing python-catch-conflicts-hook.sh
Sourcing python-remove-bin-bytecode-hook.sh
Sourcing python-imports-check-hook.sh
Using pythonImportsCheckPhase
Sourcing python-namespaces-hook
qtPreHook
unpacking sources
unpacking source archive /nix/store/v8297rqiv4wd1zyizwykwz74qbqd3dkz-QScintilla-2.11.6.tar.gz
source root is QScintilla-2.11.6
setting SOURCE_DATE_EPOCH to timestamp 1606139171 of file QScintilla-2.11.6/doc/html-Qt4Qt5/splitbar.png
patching sources
configuring
Configuring QScintilla 2.11.6...
QScintilla 2.11.6 is being used.
The QScintilla .sip files will be installed in
/nix/store/sl09w0ciwcsk74b89h1a13d5ni8dmpvz-python3.9-qscintilla-2.11.6/share/sip/PyQt5.
QScintilla will be installed in
/nix/store/sl09w0ciwcsk74b89h1a13d5ni8dmpvz-python3.9-qscintilla-2.11.6/lib/python3.9/site-packages/PyQt5.
PyQt 5.15.4 is being used.
Qt 5.12.10 is being used.
sip 4.19.25 is being used.
The sip executable is
/nix/store/n5l38hjfqhly3xzzvx83yyypfzx6lm90-python3.9-sip-4.19.25/bin/sip.
QScintilla is being built with 'protected' redefined as 'public'.
The PEP 484 stubs will be installed in
/nix/store/sl09w0ciwcsk74b89h1a13d5ni8dmpvz-python3.9-qscintilla-2.11.6/lib/python3.9/site-packages/PyQt5.
The QScintilla API file will be installed in
/nix/store/sl09w0ciwcsk74b89h1a13d5ni8dmpvz-python3.9-qscintilla-2.11.6/api/python3.9/api/python.
Generating the C++ source for the Qsci module...
Generating the .pro file for the Qsci module...
Generating the QScintilla API file...
Generating the top-level .pro file...
Generating the Makefiles...
Error: /nix/store/8mkvcnwm55ahy2lra5krhc1zxbicx6wx-qtbase-5.12.10-dev/bin/qmake
failed to create a Makefile from QScintilla.pro.
error: builder for '/nix/store/g3rv06mv76dg46gcfk80gnr89d111jhr-python3.9-qscintilla-2.11.6.drv' failed with exit code 1;
       last 10 log lines:
       > /nix/store/sl09w0ciwcsk74b89h1a13d5ni8dmpvz-python3.9-qscintilla-2.11.6/lib/python3.9/site-packages/PyQt5.
       > The QScintilla API file will be installed in
       > /nix/store/sl09w0ciwcsk74b89h1a13d5ni8dmpvz-python3.9-qscintilla-2.11.6/api/python3.9/api/python.
       > Generating the C++ source for the Qsci module...
       > Generating the .pro file for the Qsci module...
       > Generating the QScintilla API file...
       > Generating the top-level .pro file...
       > Generating the Makefiles...
       > Error: /nix/store/8mkvcnwm55ahy2lra5krhc1zxbicx6wx-qtbase-5.12.10-dev/bin/qmake
       > failed to create a Makefile from QScintilla.pro.
       For full logs, run 'nix log /nix/store/g3rv06mv76dg46gcfk80gnr89d111jhr-python3.9-qscintilla-2.11.6.drv'.
```

~~The newest version of QScintilla drops support for Qt4. Unsure how to get past this `failed to create a Makefile from QScintilla.pro` error, I have tried updating to latest qscintilla. Accordingly, this draft PR performs a few tasks, none of them successfully.~~

~~The C portion of qscintilla still builds successfully on darwin. Unfortunately, in the move to pyproject, I am having trouble getting sip configured correctly. I've tried to patch in additional PyQt5 includes into the toml file, but the build still fails with the error `AttributeError: module 'sipbuild.api' has no attribute 'prepare_metadata_for_build_wheel'`, an issue I am not sure how to resolve:~~

```
$ nix-build $NIXPKGS -A python3Packages.qscintilla-qt5
                               
this derivation will be built:
  /nix/store/61fpap2s2lq2x567awvf7sgpsf3n5kx0-python3.9-qscintilla-2.13.1.drv
building '/nix/store/61fpap2s2lq2x567awvf7sgpsf3n5kx0-python3.9-qscintilla-2.13.1.drv'...
Sourcing python-remove-tests-dir-hook
Sourcing python-catch-conflicts-hook.sh
Sourcing python-remove-bin-bytecode-hook.sh
Sourcing pip-build-hook
Using pipBuildPhase
Using pipShellHook
Sourcing pip-install-hook
Using pipInstallPhase
Sourcing python-imports-check-hook.sh
Using pythonImportsCheckPhase
Sourcing python-namespaces-hook
qtPreHook
unpacking sources
unpacking source archive /nix/store/ljszkfby50a4vzdqfhi7h4vwybjqmhzd-QScintilla_src-2.13.1.tar.gz
source root is QScintilla_src-2.13.1
setting SOURCE_DATE_EPOCH to timestamp 1634205963 of file QScintilla_src-2.13.1/doc/html/splitbar.png
patching sources
configuring
no configure script, doing nothing
building
Executing pipBuildPhase
Creating a wheel...
WARNING: The directory '/homeless-shelter/Library/Caches/pip' or its parent directory is not owned or is not writable by the current user. The cache has been disabled. Check the permissions and owner of that directory. If executing pip with sudo, you should use sudo's -H flag.
Ignoring indexes: https://pypi.org/simple
Created temporary directory: /private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/pip-ephem-wheel-cache-b0v6s6hl
Created temporary directory: /private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/pip-req-tracker-xqcpiapv
Initialized build tracking at /private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/pip-req-tracker-xqcpiapv
Created build tracker: /private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/pip-req-tracker-xqcpiapv
Entered build tracker: /private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/pip-req-tracker-xqcpiapv
Created temporary directory: /private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/pip-wheel-j3qxbh8n
Processing /private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/QScintilla_src-2.13.1/Python
  Created temporary directory: /private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/pip-req-build-a51b2ng9
  DEPRECATION: A future pip version will change local packages to be built in-place without first copying to a temporary directory. We recommend you use --use-feature=in-tree-build to test your packages with this new behavior before it becomes the default.
   pip 21.3 will remove support for this functionality. You can find discussion regarding this at https://github.com/pypa/pip/issues/7555.
  Added file:///private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/QScintilla_src-2.13.1/Python to build tracker '/private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/pip-req-tracker-xqcpiapv'
    Created temporary directory: /private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/pip-modern-metadata-vxclppiu
    Running command /nix/store/w8lqcsrwlwrvmchj46ph9880w9h4m50a-python3-3.9.6/bin/python3.9 /nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.py prepare_metadata_for_build_wheel /private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/tmporbo4upf
    Traceback (most recent call last):
      File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 126, in prepare_metadata_for_build_wheel
        hook = backend.prepare_metadata_for_build_wheel
    AttributeError: module 'sipbuild.api' has no attribute 'prepare_metadata_for_build_wheel'

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 280, in <module>
        main()
      File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 263, in main
        json_out['return_val'] = hook(**hook_input['kwargs'])
      File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 130, in prepare_metadata_for_build_wheel
        return _get_wheel_metadata_from_wheel(backend, metadata_directory,
      File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 159, in _get_wheel_metadata_from_wheel
        whl_basename = backend.build_wheel(metadata_directory, config_settings)
      File "/nix/store/drghjg1bsrxn6jc4m7n9xggk94jgb2qa-python3.9-sip-6.1.1/lib/python3.9/site-packages/sipbuild/api.py", line 51, in build_wheel
        project = AbstractProject.bootstrap('pep517')
      File "/nix/store/drghjg1bsrxn6jc4m7n9xggk94jgb2qa-python3.9-sip-6.1.1/lib/python3.9/site-packages/sipbuild/abstract_project.py", line 83, in bootstrap
        project.setup(pyproject, tool, tool_description)
      File "/nix/store/drghjg1bsrxn6jc4m7n9xggk94jgb2qa-python3.9-sip-6.1.1/lib/python3.9/site-packages/sipbuild/project.py", line 484, in setup
        self.apply_user_defaults(tool)
      File "project.py", line 47, in apply_user_defaults
        super().apply_user_defaults(tool)
      File "/nix/store/4gznpqar782bxly71y2dvqd4sd5a25rh-python3.9-pyqt-builder-1.10.1/lib/python3.9/site-packages/pyqtbuild/project.py", line 70, in apply_user_defaults
        super().apply_user_defaults(tool)
      File "/nix/store/drghjg1bsrxn6jc4m7n9xggk94jgb2qa-python3.9-sip-6.1.1/lib/python3.9/site-packages/sipbuild/project.py", line 230, in apply_user_defaults
        self.builder.apply_user_defaults(tool)
      File "/nix/store/4gznpqar782bxly71y2dvqd4sd5a25rh-python3.9-pyqt-builder-1.10.1/lib/python3.9/site-packages/pyqtbuild/builder.py", line 66, in apply_user_defaults
        raise PyProjectOptionException('qmake',
    sipbuild.pyproject.PyProjectOptionException
    Preparing wheel metadata ... error
WARNING: Discarding file:///private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/QScintilla_src-2.13.1/Python. Command errored out with exit status 1: /nix/store/w8lqcsrwlwrvmchj46ph9880w9h4m50a-python3-3.9.6/bin/python3.9 /nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.py prepare_metadata_for_build_wheel /private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/tmporbo4upf Check the logs for full command output.
ERROR: Command errored out with exit status 1: /nix/store/w8lqcsrwlwrvmchj46ph9880w9h4m50a-python3-3.9.6/bin/python3.9 /nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.py prepare_metadata_for_build_wheel /private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/tmporbo4upf Check the logs for full command output.
Exception information:
Traceback (most recent call last):
  File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_internal/cli/base_command.py", line 180, in _main
    status = self.run(options, args)
  File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_internal/cli/req_command.py", line 205, in wrapper
    return func(self, options, args)
  File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_internal/commands/wheel.py", line 142, in run
    requirement_set = resolver.resolve(
  File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/resolver.py", line 103, in resolve
    r = self.factory.make_requirement_from_install_req(
  File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/factory.py", line 444, in make_requirement_from_install_req
    raise self._build_failures[ireq.link]
  File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/factory.py", line 200, in _make_candidate_from_link
    self._link_candidate_cache[link] = LinkCandidate(
  File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/candidates.py", line 306, in __init__
    super().__init__(
  File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/candidates.py", line 151, in __init__
    self.dist = self._prepare()
  File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/candidates.py", line 234, in _prepare
    dist = self._prepare_distribution()
  File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/candidates.py", line 317, in _prepare_distribution
    return self._factory.preparer.prepare_linked_requirement(
  File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_internal/operations/prepare.py", line 508, in prepare_linked_requirement
    return self._prepare_linked_requirement(req, parallel_builds)
  File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_internal/operations/prepare.py", line 570, in _prepare_linked_requirement
    dist = _get_prepared_distribution(
  File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_internal/operations/prepare.py", line 60, in _get_prepared_distribution
    abstract_dist.prepare_distribution_metadata(finder, build_isolation)
  File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_internal/distributions/sdist.py", line 36, in prepare_distribution_metadata
    self.req.prepare_metadata()
  File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_internal/req/req_install.py", line 549, in prepare_metadata
    self.metadata_directory = self._generate_metadata()
  File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_internal/req/req_install.py", line 534, in _generate_metadata
    return generate_metadata(
  File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_internal/operations/build/metadata.py", line 31, in generate_metadata
    distinfo_dir = backend.prepare_metadata_for_build_wheel(
  File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_vendor/pep517/wrappers.py", line 184, in prepare_metadata_for_build_wheel
    return self._call_hook('prepare_metadata_for_build_wheel', {
  File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_vendor/pep517/wrappers.py", line 265, in _call_hook
    self._subprocess_runner(
  File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_internal/utils/subprocess.py", line 274, in runner
    call_subprocess(
  File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_internal/utils/subprocess.py", line 244, in call_subprocess
    raise InstallationSubprocessError(proc.returncode, command_desc)
pip._internal.exceptions.InstallationSubprocessError: Command errored out with exit status 1: /nix/store/w8lqcsrwlwrvmchj46ph9880w9h4m50a-python3-3.9.6/bin/python3.9 /nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.py prepare_metadata_for_build_wheel /private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/tmporbo4upf Check the logs for full command output.
Removed file:///private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/QScintilla_src-2.13.1/Python from build tracker '/private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/pip-req-tracker-xqcpiapv'
Removed build tracker: '/private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/pip-req-tracker-xqcpiapv'
error: builder for '/nix/store/61fpap2s2lq2x567awvf7sgpsf3n5kx0-python3.9-qscintilla-2.13.1.drv' failed with exit code 1;
       last 10 log lines:
       >     return self._call_hook('prepare_metadata_for_build_wheel', {
       >   File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_vendor/pep517/wrappers.py", line 265, in _call_hook
       >     self._subprocess_runner(
       >   File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_internal/utils/subprocess.py", line 274, in runner
       >     call_subprocess(
       >   File "/nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_internal/utils/subprocess.py", line 244, in call_subprocess
       >     raise InstallationSubprocessError(proc.returncode, command_desc)
       > pip._internal.exceptions.InstallationSubprocessError: Command errored out with exit status 1: /nix/store/w8lqcsrwlwrvmchj46ph9880w9h4m50a-python3-3.9.6/bin/python3.9 /nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.py prepare_metadata_for_build_wheel /private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/tmporbo4upf Check the logs for full command output.
       > Removed file:///private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/QScintilla_src-2.13.1/Python from build tracker '/private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/pip-req-tracker-xqcpiapv'
       > Removed build tracker: '/private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/pip-req-tracker-xqcpiapv'
       For full logs, run 'nix log /nix/store/61fpap2s2lq2x567awvf7sgpsf3n5kx0-python3.9-qscintilla-2.13.1.drv'.
```


Partially resolves: #71398

willcohen commented:
Additional googling suggests that this metadata issue means that `sip-install` needs additional flags or configuration to run successfully. Will keep looking into it. The same issue applies even when falling back to build 2.11.5 with the toml, so it’s something with my config and not with the new version.

veprbl commented on pkgs/development/python-modules/qscintilla-qt5/default.nix:
```diff
@@ -5,47 +5,51 @@
-  nativeBuildInputs = [ sip_4 qtbase ];
-  buildInputs = [ qscintilla ];
-  propagatedBuildInputs = [ pyqt5 ];
+  nativeBuildInputs = [ sip ];
```
Would `pyqt5_sip` work better here?

willcohen replied:
Is `pyqt5_sip` accessible outside of the pyqt recipe?

willcohen replied:
Apologies for my slowness -- I did see it in the `propagatedBuildInputs`, but my nix skills may be a bit lacking as this is my first PR. I can't seem to get nixpkgs to recognize `pyqt5_sip` as an available package instead, whether it's including it instead of `sip`/`pyqt5` or trying it as an attribute of `pyqt5`. Regardless, the package seems to build as noted further down, but still happy to pursue this further.

veprbl replied:
Was just a hint. Good work on this PR!

willcohen committed:
qscintilla-qt4: move qscintilla recipe here

willcohen committed:
qscintilla: update to use Qt5 only

willcohen committed:
pythonPackages.qscintilla-qt4: put qscintilla here

willcohen committed:
pythonPackages.qscintilla-qt5: use pyproject

willcohen committed:
qscintilla: default to Qt5

willcohen committed:
pythonPackages.qscintilla: default to Qt5

willcohen commented:
sip now works and is able to start the build, which now fails with:
```
    Compiling the project...
    make
    _in_process.py: 'make' failed returning 2
    cd Qsci/ && ( test -e Makefile || /nix/store/8mkvcnwm55ahy2lra5krhc1zxbicx6wx-qtbase-5.12.10-dev/bin/qmake -o Makefile /private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/tmpcvxmua7e/Qsci/Qsci.pro ) && make -f Makefile
    make[1]: Entering directory '/private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/tmpcvxmua7e/Qsci'
    clang++ -c -pipe -stdlib=libc++ -O2 -fno-exceptions -Wall -W -fPIC -DPy_LIMITED_API=0x03040000 -DSIP_PROTECTED_IS_PUBLIC -Dprotected=public -DQT_NO_EXCEPTIONS -DQT_NO_DEBUG -DQT_PLUGIN -DQT_GUI_LIB -DQT_CORE_LIB -I. -I. -I.. -isystem /nix/store/w8lqcsrwlwrvmchj46ph9880w9h4m50a-python3-3.9.6/include/python3.9 -isystem /nix/store/8mkvcnwm55ahy2lra5krhc1zxbicx6wx-qtbase-5.12.10-dev/include/QtGui -isystem /nix/store/8mkvcnwm55ahy2lra5krhc1zxbicx6wx-qtbase-5.12.10-dev/include/QtCore -I. -I/nix/store/jhj8q4vq5x3y943d8r9iyai4flsmmxch-apple-framework-OpenGL/Library/Frameworks/OpenGL.framework/Headers -I/nix/store/jwb156jli8d0236anmjlrxyflpz4ln7a-apple-framework-AGL/Library/Frameworks/AGL.framework/Headers -I/nix/store/8mkvcnwm55ahy2lra5krhc1zxbicx6wx-qtbase-5.12.10-dev/mkspecs/macx-clang -o sipQscicmodule.o sipQscicmodule.cpp
    /nix/store/imnrnq7sm16zxgxsmi9r0iqlj2jqq9c5-python3.9-PyQt5-5.15.4/lib/python3.9/site-packages/PyQt5/bindings/QtWidgets/qwidget.sip:28:10: fatal error: 'qwidget.h' file not found
    #include <qwidget.h>
             ^~~~~~~~~~~
    1 error generated.
    make[1]: *** [Makefile:553: sipQscicmodule.o] Error 1
    make[1]: Leaving directory '/private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/tmpcvxmua7e/Qsci'
    make: *** [Makefile:50: sub-Qsci-make_first-ordered] Error 2
    Preparing wheel metadata ... error
WARNING: Discarding file:///private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/QScintilla_src-2.13.1/Python. Command errored out with exit status 1: /nix/store/w8lqcsrwlwrvmchj46ph9880w9h4m50a-python3-3.9.6/bin/python3.9 /nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.py prepare_metadata_for_build_wheel /private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/tmp_hg34w_z Check the logs for full command output.
ERROR: Command errored out with exit status 1: /nix/store/w8lqcsrwlwrvmchj46ph9880w9h4m50a-python3-3.9.6/bin/python3.9 /nix/store/h5j6wymds31fgj0pdhfrdh27qhlw6d1v-python3.9-pip-21.1.3/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.py prepare_metadata_for_build_wheel /private/tmp/nix-build-python3.9-qscintilla-2.13.1.drv-0/tmp_hg34w_z Check the logs for full command output.
```

@lsix this issue with qwidget.h seems very similar to #71646, with the new issue that `configure.py` can't be patched anymore. Based on the `sip-install` help, it would seem that in addition to the settings being set to provide the locations for the various qscintilla files, that adding ` --qmake-setting 'QT += widgets' --qmake-setting 'QT += printsupport'` to the build command be an equivalent solution to what was done before, but as seen on this branch the build still fails.

```
$ nix-build $NIXPKGS --run-env -A python3Packages.qscintilla-qt5
[...]
$ sip-install --help
usage: sip-install [-h] [-V] [--quiet] [--verbose] [--build-dir DIR] [--no-distinfo] [--scripts-dir DIR] [--target-dir DIR] [--api-dir DIR] [--android-abi ABI]
                   [--link-full-dll] [--qml-debug] [--jobs N] [--qmake FILE] [--qmake-setting 'NAME += VALUE'] [--spec SPEC] [--concatenate N] [--debug] [--no-docstrings]
                   [--pep484-pyi] [--protected-is-public] [--no-protected-is-public] [--tracing] [--qsci-features-dir DIR] [--qsci-include-dir DIR] [--qsci-library-dir DIR]

Build and install the project.

optional arguments:
[...]
  --qmake-setting 'NAME += VALUE'
                        add the 'NAME += VALUE' setting to any .pro file
[...]
  --qsci-features-dir DIR
                        the qscintilla2.prf features file is in DIR
  --qsci-include-dir DIR
                        the Qsci include file directory is in DIR
  --qsci-library-dir DIR
                        the QScintilla library is in DIR
```

Additionally, I went back and tried to clean up the commit messages to get this closer to a workable PR.

willcohen committed:
pythonPackages.qscintilla-qt5: fix darwin make

willcohen committed:
pythonPackages.qscintilla-qt5: fix darwin linking

willcohen committed:
miniaudicle: specify use of qscintilla-qt4

willcohen committed:
sqliteman: specify use of qscintilla-qt4

willcohen commented:
Result of `nixpkgs-review pr 150595` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>qscintilla-qt4</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>openscad</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>qscintilla (libsForQt5.qscintilla)</li>
    <li>libsForQt514.qscintilla</li>
    <li>libsForQt515.qscintilla (plasma5Packages.qscintilla)</li>
    <li>octaveFull</li>
    <li>python38Packages.qscintilla (python38Packages.qscintilla-qt5)</li>
    <li>python39Packages.qscintilla (python39Packages.qscintilla-qt5)</li>
  </ul>
</details>

willcohen commented:
Result of `nixpkgs-review pr 150595` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>qgis</li>
    <li>qgis-unwrapped</li>
  </ul>
</details>
<details>
  <summary>11 packages built:</summary>
  <ul>
    <li>qscintilla (libsForQt5.qscintilla)</li>
    <li>libsForQt512.qscintilla</li>
    <li>libsForQt514.qscintilla</li>
    <li>octaveFull</li>
    <li>openscad</li>
    <li>ovito</li>
    <li>python38Packages.qscintilla (python38Packages.qscintilla-qt5)</li>
    <li>python39Packages.qscintilla (python39Packages.qscintilla-qt5)</li>
    <li>qscintilla-qt4</li>
    <li>tora</li>
    <li>tortoisehg</li>
  </ul>
</details>

willcohen commented:
On linux, qgis-unwrapped currently fails with:

```
error: builder for '/nix/store/gk8mdaw5h55k4yagcjp8arrb5q95xbrj-qgis-unwrapped-3.16.14.drv' failed with exit code 1;
       last 10 log lines:
       > [2780/5436] Generating gui/sip_guipart0.cpp, gui/sip_guipart1.cpp, gui/sip_guipart2.cpp, gui/sip_guipart3.cpp, gui/sip_guipart4.cpp, gui/sip_guipart5.cpp, gui/sip_guipart6.cpp, gui/sip_guipart7.cpp, gui/sip_guipart8.cpp, gui/sip_guipart9.cpp, gui/sip_guipart10.cpp/sip_corepart10.cppip_analysispart8.cpp, analysis/sip_analysispart9.cpp, analysis/sip_analysispart10.cpp
       > FAILED: python/gui/sip_guipart0.cpp python/gui/sip_guipart1.cpp python/gui/sip_guipart2.cpp python/gui/sip_guipart3.cpp python/gui/sip_guipart4.cpp python/gui/sip_guipart5.cpp python/gui/sip_guipart6.cpp python/gui/sip_guipart7.cpp python/gui/sip_guipart8.cpp python/gui/sip_guipart9.cpp python/gui/sip_guipart10.cpp /build/source/build/python/gui/sip_guipart0.cpp /build/source/build/python/gui/sip_guipart1.cpp /build/source/build/python/gui/sip_guipart2.cpp /build/source/build/python/gui/sip_guipart3.cpp /build/source/build/python/gui/sip_guipart4.cpp /build/source/build/python/gui/sip_guipart5.cpp /build/source/build/python/gui/sip_guipart6.cpp /build/source/build/python/gui/sip_guipart7.cpp /build/source/build/python/gui/sip_guipart8.cpp /build/source/build/python/gui/sip_guipart9.cpp /build/source/build/python/gui/sip_guipart10.cpp
       > cd /build/source/build/python && /nix/store/q2lj6pxd6sn75fa9pjx9x3jpgxnr6zld-cmake-3.21.2/bin/cmake -E echo && /nix/store/q2lj6pxd6sn75fa9pjx9x3jpgxnr6zld-cmake-3.21.2/bin/cmake -E touch /build/source/build/python/gui/sip_guipart0.cpp /build/source/build/python/gui/sip_guipart1.cpp /build/source/build/python/gui/sip_guipart2.cpp /build/source/build/python/gui/sip_guipart3.cpp /build/source/build/python/gui/sip_guipart4.cpp /build/source/build/python/gui/sip_guipart5.cpp /build/source/build/python/gui/sip_guipart6.cpp /build/source/build/python/gui/sip_guipart7.cpp /build/source/build/python/gui/sip_guipart8.cpp /build/source/build/python/gui/sip_guipart9.cpp /build/source/build/python/gui/sip_guipart10.cpp && /nix/store/kp9dpdpgf1gd3r9z33r7zfy49qjdxwmw-python3.9-sip-4.19.25/bin/sip -w -e -x ANDROID -x ARM -x MOBILITY_LOCATION -x VECTOR_MAPPED_TYPE -n PyQt5.sip -t Qt_5_15_0 -t WS_X11 -g -o -a /build/source/build/python/qgis.gui.api -n PyQt5.sip -y /build/source/build/output/python/qgis/_gui.pyi -I /nix/store/1dz6hmnlx15qwcqd4bvwa71qzalagzxg-python3.9-qscintilla-2.13.1/share/sip/PyQt5 -j 11 -c /build/source/build/python/gui -I /build/source/build/python/gui -I /nix/store/knynibwmvj63w0ak6w307dwd4nhcd1d7-python3.9-PyQt5-5.15.4/lib/python3.9/site-packages/PyQt5/bindings -I /build/source/python /build/source/build/python/gui/gui.sip
       >
       > sip: Unable to find file "Qsci/qscimod5.sip"
       > [2782/5436] Generating analysis/sip_analysispart0.cpp, analysis/sip_analysispart1.cpp, analysis/sip_analysispart2.cpp, analysis/sip_analysispart3.cpp, analysis/sip_analysispart4.cpp, analysis/sip_analysispart5.cpp, analysis/sip_analysispart6.cpp, analysis/sip_analysispart7.cpp, analysis/sip_analysispart8.cpp, analysis/sip_analysispart9.cpp, analysis/sip_analysispart10.cpp
       >
       > [2783/5436] Generating core/sip_corepart0.cpp, core/sip_corepart1.cpp, core/sip_corepart2.cpp, core/sip_corepart3.cpp, core/sip_corepart4.cpp, core/sip_corepart5.cpp, core/sip_corepart6.cpp, core/sip_corepart7.cpp, core/sip_corepart8.cpp, core/sip_corepart9.cpp, core/sip_corepart10.cpp
       >
       > ninja: build stopped: subcommand failed.
       For full logs, run 'nix log /nix/store/gk8mdaw5h55k4yagcjp8arrb5q95xbrj-qgis-unwrapped-3.16.14.drv'.
error: 1 dependencies of derivation '/nix/store/9z141mkxg1wdi84arxdd2yxmz474phf7-qgis-3.16.14.drv' failed to build
```

Revisions to the QGIS formula should probably be done alongside #150286

willcohen committed:
qgis: update QSCI_SIP_DIR location

veprbl commented on pkgs/top-level/all-packages.nix:
```diff
@@ -9013,7 +9013,15 @@ with pkgs;
   qprint = callPackage ../tools/text/qprint { };
 
-  qscintilla = callPackage ../development/libraries/qscintilla { };
+  qscintilla = if stdenv.isDarwin then
```
This can now migrate to `libsForQt5.qscintilla`

willcohen replied:
Thanks -- migrated accordingly.

veprbl commented on pkgs/development/libraries/qscintilla-qt4/default.nix:
```diff
@@ -0,0 +1,70 @@
+{ stdenv, lib, fetchurl, unzip
+, qt4 ? null, qmake4Hook ? null
+, withQt5 ? false, qtbase ? null, qtmacextras ? null, qmake ? null
```
This can drop the Qt5 support
```suggestion
```

willcohen replied:
I originally left this in to avoid breakage, but I think it's addressed by modifying all affected derivations still on Qt4.

veprbl commented on pkgs/development/python-modules/qscintilla-qt4/default.nix:
```diff
@@ -0,0 +1,43 @@
+
+disabledIf (isPy3k || isPyPy)
+  (buildPythonPackage {
+    # TODO: Qt5 support
```
```suggestion
```

willcohen replied:
Qt5 support dropped in the latest additional commits!

willcohen committed:
qscintilla-qt4: Remove qt5/darwin support

willcohen committed:
qscintilla: migrate to libsForQt5.qscintilla

willcohen commented:
@veprbl would it be possible to approve running ofborg on this, since it's blocked as my first-time contribution?

veprbl commented:
Is this ready for review?

willcohen commented:
Yes, this PR is ready for review! @peterhoeg @danbst @lsix @FRidh @jonringer

willcohen commented:
@veprbl is there anyone else I should be pinging on this?

willcohen commented:
@NixOS/darwin-maintainers would any of you be willing to review this? `octave` is probably the best end result example I can find that this fixes a build for on darwin.

prusnak commented on pkgs/development/libraries/qscintilla/default.nix:
```diff
@@ -1,33 +1,30 @@
 { stdenv, lib, fetchurl, unzip
-, qt4 ? null, qmake4Hook ? null
-, withQt5 ? false, qtbase ? null, qtmacextras ? null, qmake ? null
+, withQt5 ? true, qtbase, qtmacextras ? null
```
You can drop `withQt5 ? true` now as it is not used anywhere.

Also `? null` is probably not needed.

prusnak commented on pkgs/development/libraries/qscintilla/default.nix:
```diff
@@ -1,33 +1,30 @@
 let
-  pname = "qscintilla-qt${if withQt5 then "5" else "4"}";
-  version = "2.11.6";
+  pname = "qscintilla-qt5";
+  version = "2.13.1";
```
Put this few lines below instead of `inherit pname version;` and you can drop the whole `let` block.

prusnak commented on pkgs/development/libraries/qscintilla-qt4/default.nix:
```diff
@@ -0,0 +1,63 @@
+}:
+
+let
+  pname = "qscintilla-qt4";
+  version = "2.11.6";
```
Put this few lines below instead of `inherit pname version;` and you can drop the whole `let` block.

prusnak commented on pkgs/top-level/python-packages.nix:
```diff
@@ -8151,13 +8151,14 @@ in {
+  qscintilla-qt4 = callPackage ../development/python-modules/qscintilla-qt4 { };
 
   qscintilla-qt5 = pkgs.libsForQt5.callPackage ../development/python-modules/qscintilla-qt5 {
+    inherit (pkgs.qt5) qtbase;
```
Is this needed?

willcohen committed:
qscintilla, python3Packages.qscintilla-qt5: syntax

- Cleanup top-level package calls to remove unnecessary darwin
  logic and inherit statements
- Remove unnecessary let blocks in qscintilla and qscintilla-qt4
- Remove mention of withQt5

willcohen commented:
@prusnak I made all the changes accordingly (and managed to pull in a million other upstream changes when rebasing -- apologies to all for the noise.)

willcohen commented:
@prusnak the changes from your review are now just in the commit 8dfa516, and ofborg still passes all the builds. Please let me know if you see anything else!

prusnak commented on pkgs/top-level/qt5-packages.nix:
```diff
@@ -9,6 +9,7 @@
 { lib
 , pkgs
 , qt5
+, stdenv
```
This is not needed anymore.

prusnak commented on pkgs/top-level/all-packages.nix:
```diff
@@ -19492,17 +19494,17 @@ with pkgs;
 
   libsForQt515 = recurseIntoAttrs (import ./qt5-packages.nix {
-    inherit lib pkgs;
+    inherit lib pkgs stdenv;
```
This is not needed anymore.

prusnak commented on pkgs/top-level/all-packages.nix:
```diff
@@ -19492,17 +19494,17 @@ with pkgs;
 
   libsForQt514 = recurseIntoAttrs (import ./qt5-packages.nix {
-    inherit lib pkgs;
+    inherit lib pkgs stdenv;
```
This is not needed anymore.

prusnak commented on pkgs/top-level/all-packages.nix:
```diff
@@ -19492,17 +19494,17 @@ with pkgs;
 
   libsForQt512 = recurseIntoAttrs (import ./qt5-packages.nix {
-    inherit lib pkgs;
+    inherit lib pkgs stdenv;
```
This is not needed anymore.

prusnak commented on pkgs/development/python-modules/qscintilla-qt5/default.nix:
```diff
@@ -2,51 +2,53 @@
 , qscintilla
 , qtbase
+, qmake
+, qtmacextras ? null
```
```suggestion
, qtmacextras
```

prusnak commented on pkgs/development/python-modules/qscintilla-qt5/default.nix:
```diff
@@ -2,51 +2,53 @@
-  propagatedBuildInputs = [ pyqt5 ];
+  nativeBuildInputs = [ sip qmake pyqt-builder qscintilla ];
+  buildInputs = [ qtbase ];
+  propagatedBuildInputs = [ pyqt5 ] ++ lib.optional (stdenv.isDarwin) qtmacextras;
```
```suggestion
  propagatedBuildInputs = [ pyqt5 ] ++ lib.optionals stdenv.isDarwin [ qtmacextras ];
```

prusnak commented on pkgs/development/libraries/qscintilla/default.nix:
```diff
@@ -1,33 +1,27 @@
+  buildInputs = [ qtbase ];
 
-  propagatedBuildInputs = lib.optional (withQt5 && stdenv.isDarwin) qtmacextras;
+  propagatedBuildInputs = lib.optional stdenv.isDarwin qtmacextras;
```
```suggestion
  propagatedBuildInputs = lib.optionals stdenv.isDarwin [ qtmacextras ];
```

prusnak reviewed (changes requested):
Left couple more cleanup nits. Otherwise good.

willcohen committed:
qscintilla: cleanup of derivations

willcohen committed:
qt5: remove stdenv from qt5-packages

willcohen commented:
@prusnak Thanks! Comments applied with b8504e9 and a98370a.

prusnak commented:
@GrahamcOfBorg build miniaudicle qgis sqliteman python3Packages.qscintilla-qt4 python3Packages.qscintilla-qt5

siraben commented:
Result of `nixpkgs-review pr 150595` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>qscintilla</li>
  </ul>
</details>

prusnak reviewed (approved):
LGTM!

prusnak commented:
Small follow-up PR: https://github.com/NixOS/nixpkgs/pull/153744

prusnak commented on pkgs/development/python-modules/qscintilla-qt4/default.nix:
```diff
@@ -0,0 +1,42 @@
+, pyqt4
+}:
+
+disabledIf (isPy3k || isPyPy)
```
Btw, there does this limitation come from? I don't see it in the previous code

willcohen replied:
It was just a 100% copy of what was in https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/qscintilla/default.nix -- I didn't do any of the refactoring you're looking into now!

prusnak replied:
Thanks
